engine/src/main/java/org/hibernate/validator/internal/engine/validationcontext/AbstractValidationContext.java

```diff
@@ -344,27 +345,42 @@ private boolean isAlreadyValidatedForPath(Object value, PathImpl path) {
 		}
 
 		for ( PathImpl p : pathSet ) {
-			if ( path.isRootPath() || p.isRootPath() || isSubPathOf( path, p ) || isSubPathOf( p, path ) ) {
+			if ( pathsSharePrefix( p, path ) ) {
 				return true;
 			}
 		}
 
 		return false;
 	}
 
-	private boolean isSubPathOf(Path p1, Path p2) {
+	/**
+	 * this method compares two PathImpl to decide if either one is a prefix of the other.
+	 *
+	 * @param p1 first Path
+	 * @param p2 second Path
+	 * @return true if p1 ⊃ p2 or p2 ⊃ p1
+	 */
+	private boolean pathsSharePrefix(PathImpl p1, PathImpl p2) {
+		// root path is a common prefix of any path != null
+		if ( p1.isRootPath() || p2.isRootPath() ) {
+			return true;
+		}
+
+		// at this point, p1 and p2 will have at least one element in the iterator
 		Iterator<Path.Node> p1Iter = p1.iterator();
 		Iterator<Path.Node> p2Iter = p2.iterator();
 		while ( p1Iter.hasNext() ) {
 			Path.Node p1Node = p1Iter.next();
 			if ( !p2Iter.hasNext() ) {
-				return false;
+				// p1 ⊃ p2
+				return true;
 			}
 			Path.Node p2Node = p2Iter.next();
 			if ( !p1Node.equals( p2Node ) ) {
 				return false;
 			}
 		}
```

Comment:
I've taken a closer look into this. Since this `isAlreadyValidatedForPath(..)` is only applicable for an object that is about to be cascaded into, it means that the new path we are testing cannot be a root path; also since we are cascading into inner objects, i.e. going further from the object tree root, it means that the new path cannot be shorter than the ones we've already encountered. That means we can get rid of the `isRootPath()` checks here, and the only case that we are interested in should be: "is there a path in `pathSet = getInitializedProcessedPathsPerBean().get( value )` that is a prefix of the new `path` we are about to cascade into.

Then implementation-wise, I wonder if it would make more sense to move this into `PathImpl`. Since we are working with the internal `Impl` class here, we can add something like:

```java
public boolean isSubPath(PathImpl other) {
	// Ideally we would want to check this, but since we know that we are calling this only from a place where `other` is longer than `this`
	// maybe we can skip it ?
	//	if ( nodeList.size() > other.nodeList.size() ) {
	//		// cannot be a subpath as it is already longer then the other path
	//		return false;
	//	}
	for ( int i = 0; i < nodeList.size(); i++ ) {
		if ( !nodeList.get( i ).equals( other.nodeList.get( i ) ) ) {
			return false;
		}
	}
	return true;
}
```

Reply:
I agree, I also thought about moving this into PathImpl, but since the implementation is relying on the context, I was reluctant to do so. Adding the remarks from your side would reduce the scope of the method even more. Having the sentence "but since we know that we are calling this only from a place where `other` is longer than `this` maybe we can skip it ?" should not be in the public interface of PathImpl - even if it is the internal implementation. 

So, I would suggest, that a generic Method should be in PathImpl, a specific method should stay in AVC.java

I will make a new version.